lib/API/JsonApi/v1/Filters/Operators/Comparators/EqualsOperator.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ *
+ * SugarCRM Community Edition is a customer relationship management program developed by
+ * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ *
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by SalesAgility Ltd.
+ * Copyright (C) 2011 - 2017 SalesAgility Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Affero General Public License version 3 as published by the
+ * Free Software Foundation with the addition of the following permission added
+ * to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED WORK
+ * IN WHICH THE COPYRIGHT IS OWNED BY SUGARCRM, SUGARCRM DISCLAIMS THE WARRANTY
+ * OF NON INFRINGEMENT OF THIRD PARTY RIGHTS.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with
+ * this program; if not, see http://www.gnu.org/licenses or write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ *
+ * You can contact SugarCRM, Inc. headquarters at 10050 North Wolfe Road,
+ * SW2-130, Cupertino, CA 95014, USA. or at email address contact@sugarcrm.com.
+ *
+ * The interactive user interfaces in modified source and object code versions
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
+namespace SuiteCRM\API\JsonApi\v1\Filters\Operators\Comparators;
+
+use SuiteCRM\API\JsonApi\v1\Filters\Interfaces\OperatorInterface;
+use SuiteCRM\API\JsonApi\v1\Filters\Operators\Operator;
+use SuiteCRM\Exception\Exception;
+
+class EqualsOperator extends Operator implements OperatorInterface
```

Comment:
Why don't you use only one class for the Operators? Do you have any reasons for that? Because I've seen many repeating methods.

I'd use something like this

```
class Operators
{
    const OP_EQ = '=';
    const OP_GT = '>';
    const OP_GTE = '>=';
    const OP_LT = '<';
    const OP_LTE = '<=';
    const OP_NEQ = '!=';
    const OP_IS = 'IS';
    const OP_IS_NOT = 'IS NOT';
    const OP_IN = 'IN';
    ...
}

```

Reply:
Operators is the base class for the common behaviour of all operators, this is not just the SQL operators, It could represent field name comparators, and special operators. In the future we will want to be able to add special operators and perhaps future supported operations. We may even wish to support other formats like NO-SQL, JSON, XML etc.

Also each operator is responsible for its validation, operation and the operators it requires. This allows us to easily separate our concerns, so that we can have different requirements for each operators. Example: The "in" operator requires a list of operands inside brackets.

By splitting out these into classes we can use the PSR 11 container to control which operations are enabled. We can keep the complexity of the methods in the operators down.

Using your method would create a dependency, requiring the the systems which use filters to know how each operator works, which would violate the SOLID principles.